adbc.h

```diff
@@ -907,6 +1008,34 @@ AdbcStatusCode AdbcDatabaseNew(struct AdbcDatabase* database, struct AdbcError*
 AdbcStatusCode AdbcDatabaseGetOption(struct AdbcDatabase* database, const char* key,
                                      const char** value, struct AdbcError* error);
 
+/// \brief Get a bytestring option of the database.
+///
+/// This must always be thread-safe (other operations are not).
```

Comment:
Hmm... but if `AdbcDatabaseGetOptionBytes` is called from multiple threads, then it is undefined if the return `value` is still valid when the calling thread consumes it, right?

(same problem as with POSIX `getenv`, actually: https://pubs.opengroup.org/onlinepubs/9699919799/functions/getenv.html)

Reply:
Gah. Possibly "thread-safe with respect to *other* functions"? 

The only reason why I wanted it to be thread-safe was to allow for a progress indicator (on GetDouble). Maybe we can split that out into a dedicated function.

Reply:
Or you switch to malloc-allocated output, or you let the caller pass a suitable buffer area...

Reply:
So far things have avoided assuming libc malloc. We could return a struct with a release callback?

Caller-allocated buffers get annoying because you have to somehow tell the caller the size...

Reply:
> Caller-allocated buffers get annoying because you have to somehow tell the caller the size...

But at least the lifetime is safe.

Reply:
Ok, how about something like this?

```c
GetOptionBytes(..., uint8_t* value, size_t* length, ...)
/// length must be provided and must be the size of the buffer pointed
/// to by value.  If there is sufficient space, the driver will copy
/// the option value to buffer and set length to the size of the
/// actual value.  If the buffer is too small, no data will be written
/// and length will be set to the required length.
///
/// In other words:
///
/// - If output length <= input length, value will contain a value
///   with length bytes.
/// - If output length > input length, nothing has been written to
///   value.
```

Reply:
Any reason to not follow the pattern of functions like vsnprintf? i.e. print what you can but ultimately left to the caller to determine if the size was sufficient

Reply:
I was thinking it's not super useful to only get part of the input (e.g. what if it's a Protobuf? Which is a real use case with gRPC/Flight SQL).

Reply:
Frankly, I'm most inclined to revert to the original definition; multiple threads calling GetOption and invalidating the pointer is the same as a single thread calling GetOption and invalidating a previous pointer.

Reply:
`getenv` is slightly different as global state where the application may not know what libraries it uses are also doing with the global state; here the operation is scoped to a specific handle that presumably an application can manage itself

Reply:
> Ok, how about something like this?

LGTM

Reply:
Updated, I suppose we should also update the base `GetOption` (returns char*) though

Reply:
Updated that too.